# Исследование надёжности заёмщиков

<h1>Оглавление проекта<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Описание-проекта" data-toc-modified-id="Описание-проекта-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Описание проекта</a></span></li><li><span><a href="#Описание-данных" data-toc-modified-id="Описание-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Описание данных</a></span></li><li><span><a href="#Обзор-данных-" data-toc-modified-id="Обзор-данных--3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Обзор данных <a name="1_2"></a></a></span></li><li><span><a href="#Обработка-данных-" data-toc-modified-id="Обработка-данных--4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Обработка данных <a name="2"></a></a></span><ul class="toc-item"><li><span><a href="#Обработка-артифактов-в-общих-трудовых-днях-" data-toc-modified-id="Обработка-артифактов-в-общих-трудовых-днях--4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Обработка артифактов в общих трудовых днях <a name="2_1"></a></a></span></li><li><span><a href="#Обработка-артифактов-в-количестве-детей-" data-toc-modified-id="Обработка-артифактов-в-количестве-детей--4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Обработка артифактов в количестве детей <a name="2_2"></a></a></span></li><li><span><a href="#Обработка-артифактов-возраста-клиентов-" data-toc-modified-id="Обработка-артифактов-возраста-клиентов--4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>Обработка артифактов возраста клиентов <a name="2_3"></a></a></span></li><li><span><a href="#Обработка-аномалии-в-значениях-гендеров-клиентов-" data-toc-modified-id="Обработка-аномалии-в-значениях-гендеров-клиентов--4.4"><span class="toc-item-num">4.4&nbsp;&nbsp;</span>Обработка аномалии в значениях гендеров клиентов <a name="2_4"></a></a></span></li><li><span><a href="#Обработка-пропусков-" data-toc-modified-id="Обработка-пропусков--4.5"><span class="toc-item-num">4.5&nbsp;&nbsp;</span>Обработка пропусков <a name="2_5"></a></a></span></li><li><span><a href="#Изменения-типов-данных-" data-toc-modified-id="Изменения-типов-данных--4.6"><span class="toc-item-num">4.6&nbsp;&nbsp;</span>Изменения типов данных <a name="2_6"></a></a></span></li><li><span><a href="#Обработка-дублей-" data-toc-modified-id="Обработка-дублей--4.7"><span class="toc-item-num">4.7&nbsp;&nbsp;</span>Обработка дублей <a name="2_7"></a></a></span></li><li><span><a href="#Лемматизация-целей-кредита-" data-toc-modified-id="Лемматизация-целей-кредита--4.8"><span class="toc-item-num">4.8&nbsp;&nbsp;</span>Лемматизация целей кредита <a name="2_8"></a></a></span></li><li><span><a href="#Категоризация-данных-" data-toc-modified-id="Категоризация-данных--4.9"><span class="toc-item-num">4.9&nbsp;&nbsp;</span>Категоризация данных <a name="2_9"></a></a></span><ul class="toc-item"><li><span><a href="#Подготовка-сведений-о-целях-кредита-" data-toc-modified-id="Подготовка-сведений-о-целях-кредита--4.9.1"><span class="toc-item-num">4.9.1&nbsp;&nbsp;</span>Подготовка сведений о целях кредита <a name="2_9_1"></a></a></span></li><li><span><a href="#Поготовка-сведений-о-доходах-" data-toc-modified-id="Поготовка-сведений-о-доходах--4.9.2"><span class="toc-item-num">4.9.2&nbsp;&nbsp;</span>Поготовка сведений о доходах <a name="2_9_2"></a></a></span></li><li><span><a href="#Выделение-словарей-и-подготовка-набора-данных-к-анализу-" data-toc-modified-id="Выделение-словарей-и-подготовка-набора-данных-к-анализу--4.9.3"><span class="toc-item-num">4.9.3&nbsp;&nbsp;</span>Выделение словарей и подготовка набора данных к анализу <a name="2_9_3"></a></a></span></li></ul></li></ul></li><li><span><a href="#Анализ-данных-" data-toc-modified-id="Анализ-данных--5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Анализ данных <a name="3"></a></a></span><ul class="toc-item"><li><span><a href="#Связь-с-наличием-детей-" data-toc-modified-id="Связь-с-наличием-детей--5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>Связь с наличием детей <a name="3_1"></a></a></span></li><li><span><a href="#Связь-с-семейным-положением-" data-toc-modified-id="Связь-с-семейным-положением--5.2"><span class="toc-item-num">5.2&nbsp;&nbsp;</span>Связь с семейным положением <a name="3_2"></a></a></span></li><li><span><a href="#Связь-с-уровнем-дохода-" data-toc-modified-id="Связь-с-уровнем-дохода--5.3"><span class="toc-item-num">5.3&nbsp;&nbsp;</span>Связь с уровнем дохода <a name="3_3"></a></a></span></li><li><span><a href="#Связь-с-целями-" data-toc-modified-id="Связь-с-целями--5.4"><span class="toc-item-num">5.4&nbsp;&nbsp;</span>Связь с целями <a name="3_4"></a></a></span></li></ul></li><li><span><a href="#Общий-вывод-" data-toc-modified-id="Общий-вывод--6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Общий вывод <a name="4"></a></a></span></li></ul></div>

## Описание проекта

Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.
Результаты исследования будут учтены при построении модели кредитного скоринга — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

В ходе исследования будут даны ответы на следующие вопросы:
1. Есть ли зависимость между наличием детей и возвратом кредита в срок?
2. Есть ли зависимость между семейным положением и возвратом кредита в срок?
3. Есть ли зависимость между уровнем дохода и возвратом кредита в срок?
4. Как разные цели кредита влияют на его возврат в срок?

## Описание данных

- children — количество детей в семье
- days_employed — общий трудовой стаж в днях
- dob_years — возраст клиента в годах
- education — уровень образования клиента
- education_id — идентификатор уровня образования
- family_status — семейное положение
- family_status_id — идентификатор семейного положения
- gender — пол клиента
- income_type — тип занятости
- debt — имел ли задолженность по возврату кредитов
- total_income — ежемесячный доход
- purpose — цель получения кредита

## Обзор данных <a name="1_2"></a>

Познакомимся с данными.

In [1]:
import pandas as pd
data = pd.read_csv('data.csv')
data.head()

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу

Заметим, что в колонке days_employed аномальные отрицательные значения. Посмотрим на типы и пропуски. Информацию о памяти получим более полно.

In [2]:
data.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 11.4 MB


In [3]:
delta = 21525 - 19351
print(f'Строки в наборе данных: {delta}. В процентах: {delta / 21525:.0%}')

Строки в наборе данных: 2174. В процентах: 10%


Видим, что есть пропуски в колонках: days_employed, total_income. Кроме того, тип days_employed - вещественное число, тогда как стаж нам интересен с точностью до дня. Более того, переход на целочисленный тип позволит использовать меньше оперативной памяти для работы с набором данных.

## Обработка данных <a name="2"></a>

### Обработка артифактов в общих трудовых днях <a name="2_1"></a>

Теперь займемся артифактом в виде отрицательных значений по трудовому стажу. Посмотрим сколько таких записей.

In [4]:
data[data['days_employed'] < 0]['days_employed'].count()

15906

In [5]:
print(f'Доля от общего количества строк {9444 / 19351:.0%}.')

Доля от общего количества строк 49%.


Такой большой объем некорректных значений мог возникнуть в результате неправильной работы алгоритма по расчету трудовых дней. Например, расчет велся как разность двух дат: даты начала трудовой деятельности и даты завершения трудовой деятельности. Если не указать в этом расчете дату начала, только дату завершения, то как раз получатся отрицательные значения. 

Удаление значительного количества строк приведет к существенному уменьшению репрезентативности выборки. Учитывая это и то, что значения из этой колонки не будут участвовать в изучении зависимостей, то мы можем все отрицательные значения заменить на средние значения оставшейся части.

In [6]:
mean_days_employed = round(data[data['days_employed'] >= 0]['days_employed'].mean())
data.loc[data['days_employed'] < 0, 'days_employed'] = mean_days_employed

Убедимся, что все хорошо: отрицательных значений не стало, а на их место встали средние.

In [7]:
data[data['days_employed'] < 0]['days_employed'].count()

0

In [8]:
data[data['days_employed'] == mean_days_employed].head()

children  days_employed  dob_years education  education_id  \
0         1       365004.0         42    высшее             0   
1         1       365004.0         36   среднее             1   
2         0       365004.0         33   Среднее             1   
3         3       365004.0         32   среднее             1   
5         0       365004.0         27    высшее             0   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
5  гражданский брак                 1      M   компаньон     0  255763.565419   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
5               покупка жилья

Дополнительно убедимся, что у нас нет артифактов по верхним значениям. Вряд ли должно быть много записей со стажем больше 40 лет.

In [9]:
data[data['days_employed'] > 40 * 365]['days_employed'].count()

19351

Все хорошо. Их действительно нет.

### Обработка артифактов в количестве детей <a name="2_2"></a>

Посмотрим какие встречаются значения количества детей.

In [10]:
data['children'].value_counts()

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

Обращаем внимание, что у нас встречается аномальное число детей 20, причем оно ровное и без вариаций: 67 раз. Такие случаи в России в единичны, поэтому скорее всего мы имеем дело с ошибкой ввода данных: когда вводили цифру 2, то добавили лишний ноль. Поэтому мы можем заменим 20 на 2 и не потеряем важную информацию о зависимости количества детей и возвратом кредита в срок, так как общее количество таких записей в наборе мало, а вероятность встретить настоящего родителя с 20-ю детьми еще меньше.

In [11]:
data.loc[data['children'] == 20, 'children'] = 2

Кроме того, у нас встречается отрицательное значение -1: видимо в системе, откуда получено это значение, таким образом обозначается отсутствующее значение.

In [12]:
data.loc[data['children'] == -1, 'children'] = 0

Посмотрим на результаты.

In [13]:
data['children'].value_counts()

0    14196
1     4818
2     2131
3      330
4       41
5        9
Name: children, dtype: int64

Теперь значения распределены правдоподобно.

### Обработка артифактов возраста клиентов <a name="2_3"></a>

Посмотрим на распределение возрастов с шагом 10 лет.

In [14]:
def client_age_distrubution():
    ''' Процедура выводит распределение по возрасту клиентов с шагом 10 лет'''
    ages = {}
    for age in data['dob_years']:
        step = age//10
        index = f'{step}0-{step + 1}0'
        ages[index] = ages.get(index, 0) + 1
    ages = sorted(ages.items())
    print(ages)
    

client_age_distrubution()

[('00-10', 101), ('10-20', 14), ('20-30', 3169), ('30-40', 5674), ('40-50', 5371), ('50-60', 4679), ('60-70', 2346), ('70-80', 171)]


Что-то странное происходит на участке от 0 до 20 лет, среди клиентов слишком много детей. Посмотрим детальнее.

In [15]:
data[data['dob_years'] < 20]['dob_years'].value_counts()

0     101
19     14
Name: dob_years, dtype: int64

Похоже, что по 91 клиенту не указан возраст, заменим на среднее значение

In [16]:
mean_age = int(round(data[data['dob_years'] != 0]['dob_years'].mean()))
data.loc[data['dob_years'] == 0, 'dob_years'] = mean_age

Проверим результат.

In [17]:
client_age_distrubution()

[('10-20', 14), ('20-30', 3169), ('30-40', 5674), ('40-50', 5472), ('50-60', 4679), ('60-70', 2346), ('70-80', 171)]


Теперь без аномалий: возрасты указаны реалистичные.

### Обработка аномалии в значениях гендеров клиентов <a name="2_4"></a>

Посмотрим на значения

In [18]:
data['gender'].value_counts()

F      14236
M       7288
XNA        1
Name: gender, dtype: int64

XNA - аномальное значение, скорее всего вызанное сбоем системы. Посмотрим на эту строчку.

In [19]:
data[data['gender'] == 'XNA']

children  days_employed  dob_years            education  education_id  \
10701         0       365004.0         24  неоконченное высшее             2   

          family_status  family_status_id gender income_type  debt  \
10701  гражданский брак                 1    XNA   компаньон     0   

        total_income               purpose  
10701  203905.157261  покупка недвижимости

Из контекста данных восстановить значение не получается, заменим на самое популярное значение - женский пол. Эта аналитика в дальшейшем не будет использована для изучения зависимостей.

In [20]:
data.loc[data['gender'] == 'XNA', 'gender'] = 'F'

Проверим

In [21]:
data['gender'].value_counts()

F    14237
M     7288
Name: gender, dtype: int64

Теперь порядок.

### Обработка пропусков <a name="2_5"></a>

Посмотрим на пропуски.

In [22]:
data[data['days_employed'].isnull()].head()

children  days_employed  dob_years education  education_id  \
12         0            NaN         65   среднее             1   
26         0            NaN         41   среднее             1   
29         0            NaN         63   среднее             1   
41         0            NaN         50   среднее             1   
55         0            NaN         54   среднее             1   

            family_status  family_status_id gender  income_type  debt  \
12       гражданский брак                 1      M    пенсионер     0   
26        женат / замужем                 0      M  госслужащий     0   
29  Не женат / не замужем                 4      F    пенсионер     0   
41        женат / замужем                 0      F  госслужащий     0   
55       гражданский брак                 1      F    пенсионер     1   

    total_income                           purpose  
12           NaN                   сыграть свадьбу  
26           NaN                       образование  
29           NaN  строительство жилой недвижимости  
41           NaN  сделка с подержанным автомобилем  
55           NaN                   сыграть свадьбу

In [23]:
data[data['total_income'].isnull()].head()

children  days_employed  dob_years education  education_id  \
12         0            NaN         65   среднее             1   
26         0            NaN         41   среднее             1   
29         0            NaN         63   среднее             1   
41         0            NaN         50   среднее             1   
55         0            NaN         54   среднее             1   

            family_status  family_status_id gender  income_type  debt  \
12       гражданский брак                 1      M    пенсионер     0   
26        женат / замужем                 0      M  госслужащий     0   
29  Не женат / не замужем                 4      F    пенсионер     0   
41        женат / замужем                 0      F  госслужащий     0   
55       гражданский брак                 1      F    пенсионер     1   

    total_income                           purpose  
12           NaN                   сыграть свадьбу  
26           NaN                       образование  
29           NaN  строительство жилой недвижимости  
41           NaN  сделка с подержанным автомобилем  
55           NaN                   сыграть свадьбу

Похоже, что получили те же самые строки. Проверим, не связаны ли эти пропуски между собой.

In [24]:
data[data['days_employed'].isnull() & (~data['total_income'].isnull())].shape[0]

0

In [25]:
data[(~data['days_employed'].isnull()) & data['total_income'].isnull()].shape[0]

0

Обе таблицы пусты. Значит, пропуски в значениях days_employed и total_income в одних и тех же строках.

Проверим, нет ли связи между пропусками и значеними в income_type.

In [26]:
data[data['days_employed'].isnull() & data['total_income'].isnull()]['income_type'].value_counts()

сотрудник          1105
компаньон           508
пенсионер           413
госслужащий         147
предприниматель       1
Name: income_type, dtype: int64

In [27]:
data['income_type'].value_counts()

сотрудник          11119
компаньон           5085
пенсионер           3856
госслужащий         1459
безработный            2
предприниматель        2
в декрете              1
студент                1
Name: income_type, dtype: int64

Распределение значений в ограниченном наборе близко к общему набору данных - связи нет.

Кроме того, что пропуски days_employed и total_income связаны между собой, никаких зависимостей от других значений не найдено. Будем считать их случайными MAR. Такие пропуски могли появиться в результате интеграционного сбоя в системе сбора информации, часть данных могли не загрузиться.

Пропуски в колонке days_employed заменим ранее расчитанным средним значением mean_days_employed.

In [28]:
data['days_employed'] = data['days_employed'].fillna(mean_days_employed)

Для обработки пропусков в доходах без всплеска по медианному значению, получим медианы в группах по типам дохода.

In [29]:
income_medians = data[~data['total_income'].isnull()].groupby('income_type')['total_income'].median()
# переименуем колонку, чтобы потом присоединить значения к основному набору данных
income_medians = income_medians.rename('median_income')
income_medians.sort_values(ascending=False)

income_type
предприниматель    499163.144947
компаньон          172357.950966
госслужащий        150447.935283
сотрудник          142594.396847
безработный        131339.751676
пенсионер          118514.486412
студент             98201.625314
в декрете           53829.130729
Name: median_income, dtype: float64

Довольно неожиданно высокий уровень дохода у безработных, посмотрим внимательно на эти записи. 

In [30]:
data[data['income_type'] == 'безработный']

children  days_employed  dob_years education  education_id  \
3133          1  337524.466835         31   среднее             1   
14798         0  395302.838654         45    Высшее             0   

          family_status  family_status_id gender  income_type  debt  \
3133    женат / замужем                 0      M  безработный     1   
14798  гражданский брак                 1      F  безработный     0   

        total_income                  purpose  
3133    59956.991984  покупка жилья для сдачи  
14798  202722.511368             ремонт жилью

Таких клиентов всего оказалось 2 и судя по назначениям платежей: их доход связан со сдачей в аренду недвижимости. В таком случае, правильно их было бы отнести к типу дохода 'предприниматель'. Однако, медианы по этому типу мы не будем использовать в заменах пропусках или в дальнейшем анализе, поэтому можем проигнорировать эту неточность.

Присоединим медианные значения в основной набор данных, заполним пропуски в доходах медианой по группам, удалим вспомогательную колонку с медианами после использования.

In [31]:
data = data.merge(income_medians, on='income_type', how='left')
data.loc[data['total_income'].isnull(), 'total_income'] = data['median_income']
del data['median_income']

Посмотрим, что получилось.

In [32]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     21525 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      21525 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.8+ MB


Пропусков теперь нет.

### Изменения типов данных <a name="2_6"></a>

Чтобы правильно подобрать типы количественным значениям, посмотрим на максимальные-минимальные значения.

In [33]:
ds = data.describe()
ds[(ds.index == 'min')|(ds.index == 'max')]

children  days_employed  dob_years  education_id  family_status_id  debt  \
min       0.0  328728.720605       19.0           0.0               0.0   0.0   
max       5.0  401755.400475       75.0           4.0               4.0   1.0   

     total_income  
min  2.066726e+04  
max  2.265604e+06

Проверим максимальные-минимальные значения целочисленных типов в numpy (они же используются в pandas), чтобы подобрать оптимальные для нашего набора данных.

In [34]:
import numpy as np
int_types = ["uint8", "uint32"]
for it in int_types:
    print(np.iinfo(it))

Machine parameters for uint8
---------------------------------------------------------------
min = 0
max = 255
---------------------------------------------------------------

Machine parameters for uint32
---------------------------------------------------------------
min = 0
max = 4294967295
---------------------------------------------------------------



По строковым колонкам посмотрим количество различных значений, чтобы оценивать возможность измения типа на категорию.

In [35]:
str_columns = ['education', 'family_status', 'gender', 'income_type', 'purpose']
list(map(lambda name: len(data[name].value_counts()), str_columns))

[15, 5, 2, 8, 38]

Заметим, что отрицательных значений нет, учем это при выборе типа. Общий доход будем считать в целых - этого достаточно для наших аналитических целей. Остальные типы подберем по максимальным значениям.

Все строковые значения - как мы видели раньше: ограниченный набор значений, переведем в тип категории. Посмотрим на результат преобразования с подробностями по памяти.

In [36]:
data['children'] = data['children'].astype('uint8')
data['days_employed'] = data['days_employed'].astype('uint32')
data['dob_years'] = data['dob_years'].astype('uint8')
data['education'] = data['education'].astype('category')
data['education_id'] = data['education_id'].astype('uint8')
data['family_status'] = data['family_status'].astype('category')
data['family_status_id'] = data['family_status_id'].astype('uint8')
data['gender'] = data['gender'].astype('category')
data['income_type'] = data['income_type'].astype('category')
data['debt'] = data['debt'].astype('uint8')
data['total_income'] = data['total_income'].astype('uint32')
data['purpose'] = data['purpose'].astype('category')

data.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype   
---  ------            --------------  -----   
 0   children          21525 non-null  uint8   
 1   days_employed     21525 non-null  uint32  
 2   dob_years         21525 non-null  uint8   
 3   education         21525 non-null  category
 4   education_id      21525 non-null  uint8   
 5   family_status     21525 non-null  category
 6   family_status_id  21525 non-null  uint8   
 7   gender            21525 non-null  category
 8   income_type       21525 non-null  category
 9   debt              21525 non-null  uint8   
 10  total_income      21525 non-null  uint32  
 11  purpose           21525 non-null  category
dtypes: category(5), uint32(2), uint8(5)
memory usage: 1.2 MB


Действительно, типы изменились и снизился объем памяти с 11.4 MB на 1.2 MB. Разница почти на порядок.

### Обработка дублей <a name="2_7"></a>

Посмотрим какие указаны сведения об образовании.

In [37]:
data['education'].value_counts()

среднее                13750
высшее                  4718
СРЕДНЕЕ                  772
Среднее                  711
неоконченное высшее      668
ВЫСШЕЕ                   274
Высшее                   268
начальное                250
Неоконченное высшее       47
НЕОКОНЧЕННОЕ ВЫСШЕЕ       29
НАЧАЛЬНОЕ                 17
Начальное                 15
ученая степень             4
Ученая степень             1
УЧЕНАЯ СТЕПЕНЬ             1
Name: education, dtype: int64

Видим, что прежде чем провести анализ на дубли, нам необходимо привести все строковые значения к одному регистру. Пусть это будет нижний регистр. Дубли объясняются вводом данных в разных регистрах в свободной форме. Для того, чтобы избежать эту проблему в данных, следует организовать ввод сведений об образовании из фиксированного списка значений. После преобразования вернем тип к категории.

In [38]:
data['education'] = data['education'].str.lower()
data['education'] = data['education'].astype('category')

Проверим

In [39]:
data['education'].value_counts()

среднее                15233
высшее                  5260
неоконченное высшее      744
начальное                282
ученая степень             6
Name: education, dtype: int64

Теперь хорошо. Дальше проверим на полные дубли строк.

In [40]:
data.duplicated().sum()

71

Полные дубли в наборе данных отстутствуют.

### Лемматизация целей кредита <a name="2_8"></a>

Получим список лемм, указываемых в целях кредита. Посчитаем частоты полученных списков лемм.

In [41]:
from pymystem3 import Mystem
from collections import Counter

def lemmatize(row):
    global cnt
    result = m.lemmatize(row['purpose'])
    result.pop()
    i = 0
    while i < len(result):
        word = result[i]
        if not (word == ' '):
            cnt[word] += 1
            i += 1
        else:        
            result.remove(word)
    return result

cnt = Counter()
m = Mystem()

data['purpose_lemmas'] = data.apply(lemmatize, axis=1)

In [42]:
dict_cnt = dict(cnt)
dict_cnt

{'покупка': 5912,
 'жилье': 4473,
 'приобретение': 462,
 'автомобиль': 4315,
 'дополнительный': 909,
 'образование': 4022,
 'сыграть': 774,
 'свадьба': 2348,
 'операция': 2610,
 'с': 2924,
 'на': 2233,
 'проведение': 777,
 'для': 1294,
 'семья': 641,
 'недвижимость': 6367,
 'коммерческий': 1315,
 'жилой': 1233,
 'строительство': 1881,
 'собственный': 635,
 'подержать': 479,
 'свой': 2235,
 'со': 630,
 'заниматься': 908,
 'сделка': 944,
 'подержанный': 489,
 'получение': 1316,
 'высокий': 1375,
 'профильный': 436,
 'сдача': 653,
 'ремонт': 612}

In [43]:
dvs = data['purpose_lemmas'].value_counts()
dvs.index = [' '.join(purpose) for purpose in list(dvs.index)]
dvs

автомобиль                                973
свадьба                                   797
на проведение свадьба                     777
сыграть свадьба                           774
операция с недвижимость                   676
покупка коммерческий недвижимость         664
операция с жилье                          653
покупка жилье для сдача                   653
операция с коммерческий недвижимость      651
покупка жилье                             647
жилье                                     647
покупка жилье для семья                   641
строительство собственный недвижимость    635
недвижимость                              634
операция со свой недвижимость             630
строительство жилой недвижимость          626
покупка недвижимость                      624
строительство недвижимость                620
покупка свой жилье                        620
ремонт жилье                              612
покупка жилой недвижимость                607
на покупка свой автомобиль        

In [44]:
len(dvs)

37

Оставим в списке лем только существительные

In [45]:
cleared_lemmas = []
for lemma in list(dict_cnt.keys()):
    if m.analyze(lemma)[0]['analysis'][0]['gr'][0] == 'S':
        cleared_lemmas.append(lemma)
cleared_lemmas

['покупка',
 'жилье',
 'приобретение',
 'автомобиль',
 'образование',
 'свадьба',
 'операция',
 'с',
 'проведение',
 'семья',
 'недвижимость',
 'строительство',
 'сделка',
 'получение',
 'сдача',
 'ремонт']

Проясним смысл слов 'операция', 'семья' в целях

In [46]:
[purpose for purpose in dvs.index if 'операция' in purpose or 'семья' in purpose]

['операция с недвижимость',
 'операция с жилье',
 'операция с коммерческий недвижимость',
 'покупка жилье для семья',
 'операция со свой недвижимость']

Строка про семью имеет большее отношение к недвижимости, а операция не имеет отношение к медицине. Можем удалять из списка лемм. Ошибку определения части речи удалим вручную, а также отглагольные слова, которые не проясняют цель.

In [47]:
cleared_lemmas.remove('с')
cleared_lemmas.remove('покупка')
cleared_lemmas.remove('приобретение')
cleared_lemmas.remove('проведение')
cleared_lemmas.remove('получение')
cleared_lemmas.remove('сдача')
cleared_lemmas.remove('ремонт')
cleared_lemmas.remove('строительство')
cleared_lemmas.remove('сделка')
cleared_lemmas.remove('операция')
cleared_lemmas.remove('семья')

Будем назначать лемме бал за вхождение в список лемм каждой цели. Посмотрим таким образом наиболее охватывающие леммы.

In [48]:
point_cnt = Counter()
for lemma in cleared_lemmas:
    for ld_purpose in dvs.index:
        if lemma in ld_purpose:
            point_cnt[lemma] += 1
dict_points = list(dict(point_cnt).items())
dict_points.sort(key = lambda x:x[1], reverse = True)
dict_points

[('недвижимость', 10),
 ('образование', 9),
 ('автомобиль', 8),
 ('жилье', 7),
 ('свадьба', 3)]

Проанализировав частотный словарь лемм и список целей, приходим к заключению, что в качестве категорий разумно выбрать: автомобиль, свадьба, недвижимость (жилье), образование. Проверим, действительно ли каждую из целей можно таким образом классифицировать. 

In [49]:
def purpose_clasify(row):

    if 'автомобиль' in row['purpose_lemmas']:
        return 'автомобиль'
    elif 'свадьба' in row['purpose_lemmas']:
        return 'свадьба'
    elif 'образование' in row['purpose_lemmas']:
        return 'образование'
    elif ('недвижимость' in row['purpose_lemmas']) or ('жилье' in row['purpose_lemmas']):
        return 'недвижимость'
    else:
        return row['purpose_lemmas']


df = pd.DataFrame(dvs)
df = df.reset_index()
df = df.set_axis(['purpose_lemmas', 'count'], axis='columns')
df['purpose_category'] = df.apply(purpose_clasify, axis=1)
df

purpose_lemmas  count purpose_category
0                               автомобиль    973       автомобиль
1                                  свадьба    797          свадьба
2                    на проведение свадьба    777          свадьба
3                          сыграть свадьба    774          свадьба
4                  операция с недвижимость    676     недвижимость
5        покупка коммерческий недвижимость    664     недвижимость
6                         операция с жилье    653     недвижимость
7                  покупка жилье для сдача    653     недвижимость
8     операция с коммерческий недвижимость    651     недвижимость
9                            покупка жилье    647     недвижимость
10                                   жилье    647     недвижимость
11                 покупка жилье для семья    641     недвижимость
12  строительство собственный недвижимость    635     недвижимость
13                            недвижимость    634     недвижимость
14           операция со свой недвижимость    630     недвижимость
15        строительство жилой недвижимость    626     недвижимость
16                    покупка недвижимость    624     недвижимость
17              строительство недвижимость    620     недвижимость
18                      покупка свой жилье    620     недвижимость
19                            ремонт жилье    612     недвижимость
20              покупка жилой недвижимость    607     недвижимость
21              на покупка свой автомобиль    505       автомобиль
22          заниматься высокий образование    496      образование
23         сделка с подержанный автомобиль    489       автомобиль
24                         свой автомобиль    480       автомобиль
25         на покупка подержать автомобиль    479       автомобиль
26                   на покупка автомобиль    472       автомобиль
27              дополнительный образование    462      образование
28                 приобретение автомобиль    462       автомобиль
29                     сделка с автомобиль    455       автомобиль
30                     высокий образование    453      образование
31    получение дополнительный образование    447      образование
32                             образование    447      образование
33                   получение образование    443      образование
34                  профильный образование    436      образование
35           получение высокий образование    426      образование
36                  заниматься образование    412      образование

Все цели с категориями, применяем на весь набор данных.

### Категоризация данных <a name="2_9"></a>

Добавим новые колонки с категориями данных

#### Подготовка сведений о целях кредита <a name="2_9_1"></a>

Разбиваем цели кредита по категориям: автомобиль, свадьба, недвижимость (жилье), образование.

In [50]:
data['purpose_category'] = data.apply(purpose_clasify, axis=1)

data['purpose_category'].value_counts()

недвижимость    10840
автомобиль       4315
образование      4022
свадьба          2348
Name: purpose_category, dtype: int64

После категоризации уберем вспомогательные колонки из основного набора. И приведем тип к категории.

In [51]:
del data['purpose_lemmas']
data['purpose_category'] = data['purpose_category'].astype('category')
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21525 entries, 0 to 21524
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype   
---  ------            --------------  -----   
 0   children          21525 non-null  uint8   
 1   days_employed     21525 non-null  uint32  
 2   dob_years         21525 non-null  uint8   
 3   education         21525 non-null  category
 4   education_id      21525 non-null  uint8   
 5   family_status     21525 non-null  category
 6   family_status_id  21525 non-null  uint8   
 7   gender            21525 non-null  category
 8   income_type       21525 non-null  category
 9   debt              21525 non-null  uint8   
 10  total_income      21525 non-null  uint32  
 11  purpose           21525 non-null  category
 12  purpose_category  21525 non-null  category
dtypes: category(6), uint32(2), uint8(5)
memory usage: 1.2 MB


Теперь данные о целях сведены к ограниченному списку значений и готовы к дальнейшему анализу.

#### Поготовка сведений о доходах <a name="2_9_2"></a>

 Выясним диапазон значений

In [52]:
min_income = min(data['total_income'])
min_income

20667

In [53]:
max_income = max(data['total_income'])
max_income

2265604

Посмотрим на квантили по доходам, добавим начальную строчку, чтобы дальше было удобно присоединять доходы ниже квантиля 10%.

In [54]:
q_list = [.1, .3, .5, .7, .9]
q_income = data['total_income'].quantile(q_list)
q_income = q_income.astype('uint32')
q_income.index = q_income.index * 100
q_income.index = list(map(int, q_income.index.tolist()))
q_income = q_income.reset_index()
q_income = q_income.rename(columns = {'index': 'income_level', 'total_income': 'quantile'})
q_income.loc[5] = [0, min_income]
q_income = q_income.sort_values(by='income_level')
q_income = q_income.reset_index(drop = True)
q_income

income_level  quantile
0             0     20667
1            10     78811
2            30    116198
3            50    142594
4            70    179453
5            90    269568

In [55]:
def income_level(row):
    result = 0
    for q_level in q_income.itertuples():
        if row['total_income'] > q_level[2]:
            result = q_level[1]
    return result

data['income_level'] = data.apply(income_level, axis=1)

data['income_level'] = data['income_level'].astype('uint8')

data['income_level'].value_counts()

30    4447
10    4305
70    4305
50    4162
90    2153
0     2153
Name: income_level, dtype: int64

####  Выделение словарей и подготовка набора данных к анализу <a name="2_9_3"></a>

Взглянем на колонки еще раз и посмотрим, какие из них можно выделить в отдельные словари.

In [56]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21525 entries, 0 to 21524
Data columns (total 14 columns):
 #   Column            Non-Null Count  Dtype   
---  ------            --------------  -----   
 0   children          21525 non-null  uint8   
 1   days_employed     21525 non-null  uint32  
 2   dob_years         21525 non-null  uint8   
 3   education         21525 non-null  category
 4   education_id      21525 non-null  uint8   
 5   family_status     21525 non-null  category
 6   family_status_id  21525 non-null  uint8   
 7   gender            21525 non-null  category
 8   income_type       21525 non-null  category
 9   debt              21525 non-null  uint8   
 10  total_income      21525 non-null  uint32  
 11  purpose           21525 non-null  category
 12  purpose_category  21525 non-null  category
 13  income_level      21525 non-null  uint8   
dtypes: category(6), uint32(2), uint8(6)
memory usage: 1.2 MB


Выделим сведения об образовании в отдельный набор данных, устраним дубли.

In [57]:
education_dict = data[['education_id','education']]
education_dict = education_dict.drop_duplicates().reset_index(drop=True)
education_dict

education_id            education
0             0               высшее
1             1              среднее
2             2  неоконченное высшее
3             3            начальное
4             4       ученая степень

Выделим сведения об семейном положении в отдельный набор данных, устраним дубли.

In [58]:
family_status_dict = data[['family_status_id','family_status']]
family_status_dict = family_status_dict.drop_duplicates().reset_index(drop=True)
family_status_dict

family_status_id          family_status
0                 0        женат / замужем
1                 1       гражданский брак
2                 2         вдовец / вдова
3                 3              в разводе
4                 4  Не женат / не замужем

Аналогично выделим сведения о типе дохода в отдельный словарь.

In [59]:
income_type_dict = data[['income_type']]
income_type_dict = income_type_dict.drop_duplicates().reset_index(drop=True)
income_type_dict

income_type
0        сотрудник
1        пенсионер
2        компаньон
3      госслужащий
4      безработный
5  предприниматель
6          студент
7        в декрете

Выделим категорию целей в словарь и добавим идентификатор категории цели в первоначальный набор данных.

In [60]:
purpose_category_dict = data[['purpose_category']]
purpose_category_dict = purpose_category_dict.drop_duplicates().reset_index(drop=True)
purpose_category_dict['purpose_category_id'] = purpose_category_dict.index
data_purpose_subcategory = data.merge(purpose_category_dict, on='purpose_category', how='left')
data['purpose_category_id'] = data_purpose_subcategory['purpose_category_id']
data['purpose_category_id'] = data['purpose_category_id'].astype('uint8')
purpose_category_dict

purpose_category  purpose_category_id
0     недвижимость                    0
1       автомобиль                    1
2      образование                    2
3          свадьба                    3

Поготовим набор данных, который бы состоял только из категорий необходимых для анализа, уже приведенные к идентификаторам.

In [61]:
prepared_data = data[['children', 'family_status_id', 'income_level', 'purpose_category_id', 'debt']]
prepared_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21525 entries, 0 to 21524
Data columns (total 5 columns):
 #   Column               Non-Null Count  Dtype
---  ------               --------------  -----
 0   children             21525 non-null  uint8
 1   family_status_id     21525 non-null  uint8
 2   income_level         21525 non-null  uint8
 3   purpose_category_id  21525 non-null  uint8
 4   debt                 21525 non-null  uint8
dtypes: uint8(5)
memory usage: 913.3 KB


## Анализ данных <a name="3"></a>

### Связь с наличием детей <a name="3_1"></a>

Агрегируем данные по проблемам с возвратом в зависимости от количества детей.

In [62]:
aggegated_data = prepared_data.groupby('children').agg({'debt':['sum', 'count']})
aggegated_data

debt       
             sum  count
children               
0         1064.0  14196
1          444.0   4818
2          202.0   2131
3           27.0    330
4            4.0     41
5            0.0      9

Видим, что чем меньше детей, тем больше проблем с возвратами кредитов. Однако, не будем торопиться с выводами и учтем, что бездетных и малодетных клиентов больше по количеству. Посчитаем доли проблемных клиентов по каждому количеству детей. 

In [63]:
aggegated_data['problem_ratio'] = aggegated_data['debt']['sum']/aggegated_data['debt']['count']
aggegated_data['problem_ratio'] = round(100 * aggegated_data['problem_ratio'], 1)
aggegated_data

debt        problem_ratio
             sum  count              
children                             
0         1064.0  14196           7.5
1          444.0   4818           9.2
2          202.0   2131           9.5
3           27.0    330           8.2
4            4.0     41           9.8
5            0.0      9           0.0

В действительности, доли близки по значениям для каждого количества детей. Для 5-ти детей тоже вывод сделать не можем, потому что таких клиентов было 8 человек всего.

Чтобы изучить как вообще влияет наличие детей на надежность заемщиков, выделим признак наличия детей в отдельную колонку.

In [64]:
aggegated_data['has_children'] = aggegated_data.index > 0
aggegated_data = aggegated_data.groupby('has_children').sum()
aggegated_data['problem_ratio'] = aggegated_data['debt']['sum']/aggegated_data['debt']['count']
aggegated_data['problem_ratio'] = round(100 * aggegated_data['problem_ratio'], 1)
aggegated_data

debt        problem_ratio
                 sum  count              
has_children                             
False         1064.0  14196           7.5
True           677.0   7329           9.2

Клиенты с детьми имеют немного больше проблем с задолженностями: 9% против 7% бездетных.

Таким образом, мы можем сделать вывод, что наличие детей осложняет возврат кредитов, в связи с общем усложнением жизни заемщиков, но не намного, так как дети также повышают ответственность клиентов. 

### Связь с семейным положением <a name="3_2"></a>

Проанализируем надежность клиентов в зависимости от семейного положения. Для этого сгрупируем подготовленные данные по семейному статусу и посчитаем доли клиентов, имевших проблемы с кредитами, для каждого семейного статуса.

In [65]:
aggegated_data = prepared_data.groupby('family_status_id').agg({'debt':['sum', 'count']})
aggegated_data['problem_ratio'] = aggegated_data['debt']['sum']/aggegated_data['debt']['count']
# двухуровневые колонки заменим на одноуровневые, чтобы не было проблем при слиянии
aggegated_data.columns = aggegated_data.columns.map('_'.join).str.strip('_')
aggegated_data = aggegated_data.merge(family_status_dict, on='family_status_id', how='left')
aggegated_data['problem_ratio'] = round(100 * aggegated_data['problem_ratio'], 1)
aggegated_data.sort_values(by = 'problem_ratio')

family_status_id  debt_sum  debt_count  problem_ratio  \
2                 2      63.0         960            6.6   
3                 3      85.0        1195            7.1   
0                 0     931.0       12380            7.5   
1                 1     388.0        4177            9.3   
4                 4     274.0        2813            9.7   

           family_status  
2         вдовец / вдова  
3              в разводе  
0        женат / замужем  
1       гражданский брак  
4  Не женат / не замужем

Видим, что больше всего проблем у людей, необремененных браком ни в текущем, ни в прошлом. Те, кто состоит в текущем браке относятся к финансовым вопросам ответственнее. Самыми ответственными оказались те, для кого брак в прошлом.

Такие результаты объясняются тем, что брак приучает людей к ответственности.

### Связь с уровнем дохода <a name="3_3"></a>

In [66]:
aggegated_data = prepared_data.groupby('income_level').agg({'debt':['sum', 'count']})
aggegated_data['problem_ratio'] = aggegated_data['debt']['sum']/aggegated_data['debt']['count']
# двухуровневые колонки заменим на одноуровневые, чтобы не было проблем при слиянии
aggegated_data.columns = aggegated_data.columns.map('_'.join).str.strip('_')
aggegated_data = aggegated_data.merge(q_income, on='income_level', how='left')
aggegated_data['problem_ratio'] = round(100 * aggegated_data['problem_ratio'], 1)
aggegated_data['quantile'] = round(aggegated_data['quantile'], -4)
aggegated_data = aggegated_data.rename(columns={'quantile':'income'})
aggegated_data.sort_values(by = 'problem_ratio')

income_level  debt_sum  debt_count  problem_ratio  income
5            90     151.0        2153            7.0  270000
0             0     158.0        2153            7.3   20000
4            70     332.0        4305            7.7  180000
3            50     348.0        4162            8.4  140000
1            10     366.0        4305            8.5   80000
2            30     386.0        4447            8.7  120000

Самыми проблемными оказались клиенты со средним достатком 80 - 180 тыс. рублей. Люди с низкими и высокими доходами более ответственно относились к своим кредитным обязательствам. Меньше всего проблем было с людьми, чей доход выше 270 тыс. рублей.

Видимо, это объясняется психологическими причинами. Богатые и бедные объективнее оценивают свои возможности, а у среднего класса больше искушений жить не по средствам.

### Связь с целями <a name="3_4"></a>

Аналогично изучим влияние целей кредитов на способность возвращать деньги.

In [67]:
aggegated_data = prepared_data.groupby('purpose_category_id').agg({'debt':['sum', 'count']})
aggegated_data['problem_ratio'] = aggegated_data['debt']['sum']/aggegated_data['debt']['count']
# двухуровневые колонки заменим на одноуровневые, чтобы не было проблем при слиянии
aggegated_data.columns = aggegated_data.columns.map('_'.join).str.strip('_')
aggegated_data = aggegated_data.merge(purpose_category_dict, on='purpose_category_id', how='left')
aggegated_data['problem_ratio'] = round(100 * aggegated_data['problem_ratio'], 1)
aggegated_data.sort_values(by = 'problem_ratio')

purpose_category_id  debt_sum  debt_count  problem_ratio purpose_category
0                    0     782.0       10840            7.2     недвижимость
3                    3     186.0        2348            7.9          свадьба
2                    2     370.0        4022            9.2      образование
1                    1     403.0        4315            9.3       автомобиль

Результаты показывают, что при заеме на образование или автомобиль, возникает чуть больше проблем, чем при приобретении недвижимость или на свадьбу. Объясняться это может тем, что недвижимость и брак дают клиентам больше стабильности.

## Общий вывод <a name="4"></a>

Проведенное исследование показывает, что клиенты с детьми, несостоящие в браке, со средним доходом (80-180 тыс. рублей), берущие деньги на автомобиль или образование, чаще всего имеют проблемы с возратом кредитов.

Эти факторы нужно учесть при построении модели кредитного скоринга.
